### PR TITLE
feat: adds allowed tokens list

### PR DESCRIPTION
**Pull request by 0xged**
In this PR we add the allow list functionality to:

- deposit
- increasePosition

and all the management functions.

### CODE REVIEW

**Comment by 0xged** [comment]
Also, there is an "optimization" where we can remove the `_storeMagnitudeIfNecessary` from `deposits` and move it to the moment where we allow the tokens.

**Comment by 0xsambugs** [comment]
> Obviously a draft. Will work on tests once we agree with the scope of the feature.
> 
> @0xsambugs wdyt? I'm leaving the possibility to "unallow" a token, after it was allowed basically because if a token is upgradeable and they do weird stuff we need to be able to stop interactions with those tokens, BUT we will ALWAYS let our users withdraw tokens. No matter what.

1. Overall looks good and makes sense
2. I would say that we can avoid the `_assertNonZeroAddress` check. This is something that will be only called by us, and I don't think the zero address can break anything
3. Do you think it still makes sense to store the magnitude, or should we go back to calling decimals? I mean, since we will be having an allow list, we are not vulnerable anymore. I'm not sure what the best options is, just wanted to see what you think
4. If I understand correctly, "unallowing" a token would prevent further deposits, right? Would we still allow swaps? If not, then will we need to check each address on a `swap`?

**Comment by 0xsambugs** [comment]
3. Since we can allow/disallow tokens, I think we should do whatever is simpler. And I believe that querying the token's decimals is simpler than storing/caching them
4.B. I meant that adding the check in swaps will add an extra cost. I would like to understand how big that extra cost is

File: contracts/DCAHub/DCAHubConfigHandler.sol [hunk 1]
```diff
@@ -29,6 +29,8 @@ abstract contract DCAHubConfigHandler is DCAHubParameters, AccessControl, Pausab
   /// @inheritdoc IDCAHubConfigHandler
   uint16 public platformFeeRatio = 2500; // 25%
 
+  mapping(address => bool) internal _allowedTokens;
```
Comment (nchamo) [inline contracts/DCAHub/DCAHubConfigHandler.sol:32]: Why internal? It woud be nice to be able to query this from etherscan for example

File: contracts/DCAHub/DCAHubPositionHandler.sol [hunk 2]
```diff
@@ -196,6 +197,8 @@ abstract contract DCAHubPositionHandler is ReentrancyGuard, DCAHubConfigHandler,
       _increase ? IDCAPermissionManager.Permission.INCREASE : IDCAPermissionManager.Permission.REDUCE
     );
 
+    if (_increase && (!_allowedTokens[_userPosition.from] || !_allowedTokens[_userPosition.to])) revert IDCAHubConfigHandler.UnallowedToken();
```
Comment (nchamo) [inline contracts/DCAHub/DCAHubPositionHandler.sol:200]: I think this should be in `increasePosition` instead of here

**Review by nchamo — COMMENTED** [review]
Great work, just made two comments

File: contracts/DCAHub/DCAHubPositionHandler.sol (hunk 2 above)
Comment (0xsambugs) [reply contracts/DCAHub/DCAHubPositionHandler.sol:200]: Nit: maybe we could create an internal function that checks 
```if (!_allowedTokens[_from] || !_allowedTokens[_to]) revert IDCAHubConfigHandler.UnallowedToken();```

**Review by 0xsambugs — APPROVED** [review]

File: contracts/DCAHub/DCAHubPositionHandler.sol (hunk 2 above)
Comment (0xged) [reply contracts/DCAHub/DCAHubPositionHandler.sol:199]: @nchamo the thing is that if I check this in `increasePosition` I will have to access to user position's through storage => costly.

Comment (nchamo) [reply contracts/DCAHub/DCAHubPositionHandler.sol]: Oh, I see. Ok, makes sense then

File: test/unit/DCAHub/dca-hub-config-handler.spec.ts [hunk 3]
```diff
@@ -105,6 +108,44 @@ contract('DCAHubConfigHandler', () => {
...
+      then('tx is reverted with reason', async () => {
+        await behaviours.txShouldRevertWithMessage({
+          contract: DCAHubConfigHandler,
+          func: 'setAllowedTokens',
+          args: [[generateRandomAddress()], []],
+          message: 'InvalidAllowedTokensInput',
+        });
+      });
+    });
+    when('setting allowed tokens for the first time', () => {
```
[the remark this answers is not in the document]
Comment (0xged) [reply test/unit/DCAHub/dca-hub-config-handler.spec.ts:129]: Yeap. Also added another e2e unallowing tokens + allowing them and testing swap & terminate after that.

**Pull request merged**